doc/BorderRouter.md

```diff
@@ -0,0 +1,124 @@
+The current design was not created with performance in mind.
+Previously, there also existed a design which did something similar than the proposed design in this document.
+[See](https://github.com/scionproto/scion/tree/92531f5cb62197b9d705001c13e5a6bdb7ba1fa4/go/border).
+
+# Overview
+The pipeline gets changed to have seperate goroutines for the receiving, processing and forwarding steps.
+This will lead to a much higher performance because the expensive processing logic is moved to other
+goroutines and to improve the performance we can just increase the number of processing routines.
+
+# Design
+The border router will consist of three layers, the receiving, the processing and the forwarding layer.
+The communication between those layers are implemented as go channels.
+
+* **Receivers** There is one receiver per border router interface that is responsible for batch-reading the
+packets from the network socket, identifying the source and flowID and using them to identify which processing
+routine has to process the packet.
+Then the receiver enqueues the packet to that processing routine.
+If the queue of that processing routine is full, the packet will be dropped.
+Each receiver has a pool of preallocated packet buffers that they can use to store the packets they receive.
+This can be implemented as a go channel of packets where the receiver reads a certain amount of packets,
+updates the pointers of the ipv4.Message.Buffers to the buffer inside those packets and then performs a batch read.
+* **Processing Routines** There are several processing routines and slow-path processing routines
+in the border router that are responsible for processing the received packet.
+The actual processing logic remains unchanged.
+If the processing routine identifies a packet that belongs to the slow-path, the processing routines enqueues
+the packet to a slow-path processing routine. If the queue of the slow-path processing routine is full, the
+packet will not be processed at all. In this case the buffer is immediately returned to the receiver.
+Once a packet is processed, it gets enqueued to the forwarder which is responsible for the egress interface.
+If the queue of the forwarder is full, the packet will be dropped.
+* **Forwarders** There exists one forwarder per border router interface that is responsible for forwarding the
+packets over the network that it receives from the processing routines. It forwards the packets as a batch.
+Afterwards it returns each buffer to the receiver from which it originates.
+
+![Border Router Design](fig/border_router/br_design.png)
+
+## Mapping of processing routines
+To prevent any packet reordering on the fast-path, we map the tuple of source and flowID, see the
+[SCION header documentation](https://github.com/scionproto/scion/blob/master/doc/protocols/scion-header.rst),
+to a fixed processing routine using a hash function together with a secret value to prevent pre-computations
```

Comment:
It it is a _secret value_, we have to care about things like key rotation.
If this is instead a _salt_ or _random value_, we only have to keep it in memory.

Reply:
My idea was to use a salt but Marc mentioned something about secrets, so I am not sure about that.